cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/cloudstorage/coordinated/TwoPhaseImportCoordinator.java

```diff
@@ -0,0 +1,267 @@
+
+    private final long startTimeNanos;
+    private final CoordinatedCloudStorageDataTransferApi dataTransferApi;
+    private final JobInfo job;
+    private final StorageTransportExtension extension;
+    private final CompletableFuture<RestoreJobStatus> stageReady = new CompletableFuture<>();
+    private final CompletableFuture<RestoreJobStatus> importReady = new CompletableFuture<>();
+    private final Set<String> stageCompletedClusters;
+    private final Set<String> importCompletedClusters;
+
+    private volatile boolean succeeded = false;
+    private volatile ImportFailedException importFailedException = null;
+
+    @VisibleForTesting
+    TwoPhaseImportCoordinator(long startTimeNanos,
+                              JobInfo job,
+                              CoordinatedCloudStorageDataTransferApi dataTransferApi,
+                              StorageTransportExtension extension)
+    {
+        this.startTimeNanos = startTimeNanos;
+        this.job = job;
+        this.dataTransferApi = dataTransferApi;
+        this.stageCompletedClusters = ConcurrentHashMap.newKeySet(dataTransferApi.size());
+        this.importCompletedClusters = ConcurrentHashMap.newKeySet(dataTransferApi.size());
+        this.extension = extension;
+    }
+
+    public static TwoPhaseImportCoordinator of(long startTimeNanos,
+                                               JobInfo job,
+                                               CoordinatedCloudStorageDataTransferApi dataTransferApi,
+                                               StorageTransportExtension extension)
+    {
+        return new TwoPhaseImportCoordinator(startTimeNanos,
+                                             job,
+                                             dataTransferApi,
+                                             extension);
+    }
+
+    @Override
+    public boolean succeeded()
+    {
+        return succeeded;
+    }
+
+    @Override
+    public ImportFailedException importFailure()
+    {
+        return importFailedException;
+    }
+
+    @Override
+    public void await() throws ImportFailedException
+    {
+        try
+        {
+            awaitInternal();
+            succeeded = true;
+        }
+        catch (Exception cause)
+        {
+            importFailedException = new ImportFailedException(cause);
+            throw importFailedException;
+        }
+    }
+
+    @Override
+    public void onStageReady(String jobId)
+    {
+        validateReceivedJobId(jobId, job);
+        LOGGER.info("Received StageReady signal for coordinated write. Notifying Sidecars. jobId={}", jobId);
+        stageReady.complete(RestoreJobStatus.STAGE_READY);
+    }
+
+    @Override
+    public void onApplyReady(String jobId)
+    {
+        validateReceivedJobId(jobId, job);
+        LOGGER.info("Received ApplyReady signal for coordinated write. Notifying Sidecars. jobId={}", jobId);
+        importReady.complete(RestoreJobStatus.IMPORT_READY);
```

Comment:
Can we be consistent with the Stage/Apply terminology here?

Reply:
ok.